File: src/clients.rs
use crate::services::{Service, ServiceBucket};
use crate::utils::RandTable;
use std::fmt;

pub struct Client {
    pub id: usize,
    pub priority: usize,
    pub elapsed_time: i128,
    pub service: Service,
}

pub fn populate_clients(
    num: usize,
    out: &mut Vec<Client>,
    service_bucket: &mut ServiceBucket,
    priority_table: &mut RandTable<usize>,
) {
    for id in 0..num {
        out.push(Client {
            id: id,
            priority: priority_table.random(),
            elapsed_time: 0,
            service: service_bucket.random(),
        });
    }
}

impl fmt::Display for Client {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{};{};{};{}",
            self.id, self.service, self.priority, self.elapsed_time
        )
    }
}

File: c++/clients.cpp
//
// Created by Tommy Pujol on 4/19/21.
//

#include "clients.h"

std::string Client::to_string() {
    return boost::str(boost::format("%d;%s;%d;%d") % this->id % this->service->name % this->priority % Time::human_time(this->elapsed_time));
}

namespace Core {
    void populate_clients(int num, vector<Client> &out, ServiceBucket service_bucket, RandTable<int> &priority_table) {
        for (int i = 0; i != num; i++) {
            out.emplace_back(Client(
                    i, *priority_table.random(), Time::none(), service_bucket.random()
            ));
        }
    }
}